frame/merkle-mountain-range/rpc/src/lib.rs

```diff
@@ -258,6 +250,21 @@ fn mmr_error_into_rpc_error(err: MmrError) -> CallError {
 			"Error while generating the proof",
 			Some(data),
 		)),
+		MmrError::Verify => CallError::Custom(ErrorObject::owned(
+			VERIFY_PROOF_ERROR,
+			"Error while verifying the proof",
+			Some(data),
+		)),
+		MmrError::BlockNumToLeafIndex => CallError::Custom(ErrorObject::owned(
+			BLOCK_NUM_TO_LEAF_INDEX_ERROR,
+			"Error while converting block number to leaf index",
+			Some(data),
+		)),
+		MmrError::InvalidBestKnownBlock => CallError::Custom(ErrorObject::owned(
+			INVALID_BEST_KNOWN_BLOCK_ERROR,
+			"Invalid best known block",
+			Some(data),
+		)),
```

Comment:
thanks for your detailed take! I've gone with your suggestion to inline using `thiserror` in https://github.com/paritytech/substrate/pull/12530/commits/d5f4cf7de571e11701aeb8738b79df215cd916d4, so error msgs are no longer divorced from the error enum.

> If it makes sense to return CallError::Failed.

We lose error code granularity doing that, and in substrate, `CallError::Failed` is otherwise only used in https://github.com/paritytech/substrate/blob/63d76941b0b96df31d8c11930a0149fd5d5dda9a/client/rpc-api/src/author/error.rs#L179-L180 for handling rpc policy denials, so I've remained with `CallError::Custom`.
The error codes are still defined in `pallet-mmr-rpc`.

Reply:
Looks good. Thanks !